File: src/qda.c
// qda.c - Quadratic Discriminant Analysis routines
//
// by Artem Sokolov

#include <stdio.h>

#include "matrix.h"

// Trains a QDA classifier
//
// Inputs:
// 	i_x - N x p data matrix of N samples in p dimensions
// 	i_y - N x 1 vector of labels
//	i_N - number of samples
//	i_p - data dimensionality
//	i_K - number of classes
//	i_debug_output - level of debug output
//
// Output:
//	o_priors - K x 1 vector of Bayesian priors, computed as fraction of points from each class
//	o_means - K x p matrix of means approximated from the data
//	o_covmats - K x p x p array of covariance matrices approximated from the data
//	o_icovmats - K x p x p array of inverse covariance matrices
//	o_bias - K x 1 vector of bias terms for discriminant function computations
//
// Memory allocation requirements for outputs:
//	o_priors must be of length K
//	o_means must be of length K*p
//	o_covmat must be of length K*p*p
//	o_icovmat must be of length K*p*p
//	o_bias must be of lenght K
//	o_status must be of length 1
//	o_info must be of length 1
//
// Meaning of o_status:
//	0 - Everything is OK
//	1 - The covariance matrix for class *o_info is singular
void QDAtrain( 
	      double* i_x, int* i_y, int* i_N, int* i_p, int* i_K, double* i_cov_reg,
	      int* i_debug_output,
	      double* o_priors, double* o_means, double* o_covmats, double* o_icovmats,
	      double* o_bias, int* o_status, int* o_info
	      )
{
  if( *i_debug_output > 0 )
    printf( "Currently running C version of the QDA.train code\n" );

  // Create local storage
  int i, j, t, k;
  int* Nk = (int*)malloc( *i_K * sizeof(int) );
  double** x = make2D( *i_N, *i_p );
  double** means = make2D( *i_K, *i_p );
  double*** covmats = make3D( *i_K, *i_p, *i_p );
  double*** icovmats = make3D( *i_K, *i_p, *i_p );

  // Copy and reshape the inputs
  cp_vec_mat( *i_N, *i_p, i_x, x );

  // Init the counts and means to 0
  for( k = 0; k < *i_K; k++ )
    {
      Nk[k] = 0;
      for( i = 0; i < *i_p; i++ )
	means[k][i] = 0.0;
    }

  // Init the covariance matrices to 0
  for( k = 0; k < *i_K; k++ )
    for( i = 0; i < *i_p; i++ )
      for( j = 0; j < *i_p; j++ )
	covmats[k][i][j] = 0.0;

  // Traverse one sample at a time to compute mean contributions
  for( t = 0; t < *i_N; t++ )
    {
      k = i_y[t] - 1;		// account for 0-based index

      // Count the sample
      Nk[k]++;

      // Add contribution to the mean
      add_vec_vec( *i_p, 1.0, 1.0, means[k], x[t], means[k] );
    }

  // Compute the means and the priors
  for( k = 0; k < *i_K; k++ )
    {
      o_priors[k] = ((double)Nk[k]) / ((double)*i_N);
      mult_const_vec( *i_p, 1.0 / ((double)Nk[k]), means[k] );
    }

  // Subtract the means from the samples and compute the covariance matrices
  for( t = 0; t < *i_N; t++ )
    {
      k = i_y[t] - 1;		// account for 0-based index
      add_vec_vec( *i_p, 1.0, -1.0, x[t], means[k], x[t] );

      for( i = 0; i < *i_p; i++ )
	for( j = 0; j < *i_p; j++ )
	  covmats[k][i][j] += x[t][i] * x[t][j];
    }

  for( k = 0; k < *i_K; k++ )
    for( i = 0; i < *i_p; i++ )
      for( j = 0; j < *i_p; j++ )
	covmats[k][i][j] /= (Nk[k] - 1);

  // Regularize the covariance matrices
  if( *i_cov_reg > 0 )
    {
      for( k = 0; k < *i_K; k++ )
	for( i = 0; i < *i_p; i++ )
	  for( j = 0; j < *i_p; j++ )
	    {
	      covmats[k][i][j] *= (1 - *i_cov_reg);
	      if( i == j )
		covmats[k][i][j] += *i_cov_reg;
	    }
    }
  
  // Compute the bias terms
  for( k = 0; k < *i_K; k++ )
    o_bias[k] = log( o_priors[k] ) - 0.5 * covmat_logdet( *i_p, covmats[k] );

  // Compute the covariance matrix inverses
  for( k = 0; k < *i_K; k++ )
    {
      *o_status = covmat_inverse( *i_p, covmats[k], icovmats[k] );
      if( *o_status != 0 )
	{
	  free2D( x, *i_N );
	  free2D( means, *i_K );
	  free3D( covmats, *i_K, *i_p );
	  free3D( icovmats, *i_K, *i_p );
	  *o_info = k;
	  return;
	}
    }

  // Copy the results from local storage to outputs
  cp_mat_vec( *i_K, *i_p, means, o_means );
  cp_3D_vec( *i_K, *i_p, *i_p, covmats, o_covmats );
  cp_3D_vec( *i_K, *i_p, *i_p, icovmats, o_icovmats );

  // Free memory
  free2D( x, *i_N );
  free2D( means, *i_K );
  free3D( covmats, *i_K, *i_p );
  free3D( icovmats, *i_K, *i_p );
}

// Applies previously trained QDA classifier to new data matrix x
//
// Inputs:
// 	i_x - N x p data matrix of N samples in p dimensions
//	i_means - K x p matrix of class centroids
// 	i_icovmats - K x p x p array of inverse covariance matrices
//	i_bias - vector of length containing the bias term for each discriminant function
//	i_N - number of samples in x
//	i_p - dimensionality of x
//	i_K - number of classes
//
// Outputs:
//	o_labels - labels for all samples in i_x
//
// Memory allocation requirements for the outputs:
//	o_labels must be of length N
void QDAtest(
	     double* i_x, double* i_means, double* i_icovmats, double* i_bias,
	     int* i_N, int* i_p, int* i_K, int* o_labels
	     )
{
  // Create local storage
  int t, k, i, j;
  double temp;
  double* x_c = make1D( *i_p );
  double** x = make2D( *i_N, *i_p );
  double** means = make2D( *i_K, *i_p );
  double*** icovmats = make3D( *i_K, *i_p, *i_p );
  double** delta = make2D( *i_N, *i_K );

  // Copy and reshape the inputs to local storage
  cp_vec_mat( *i_N, *i_p, i_x, x );
  cp_vec_mat( *i_K, *i_p, i_means, means );
  cp_vec_3D( *i_K, *i_p, *i_p, i_icovmats, icovmats );

  // Compute the delta function values
  for( t = 0; t < *i_N; t++ )
    for( k = 0; k < *i_K; k++ )
      {
	add_vec_vec( *i_p, 1.0, -1.0, x[t], means[k], x_c );
	delta[t][k] = 0.0;
	for( i = 0; i < *i_p; i++ )
	  for( j = 0; j < *i_p; j++ )
	    delta[t][k] += icovmats[k][i][j] * x_c[i] * x_c[j];
	delta[t][k] *= -0.5;
	delta[t][k] += i_bias[k];
      }

  // Pick the highest discriminant function value for each point
  for( t = 0; t < *i_N; t++ )
    {
      o_labels[t] = 1;
      temp = delta[t][0];
      for( k = 1; k < *i_K; k++ )
	{
	  if( delta[t][k] > temp )
	    {
	      o_labels[t] = k + 1;
	      temp = delta[t][k];
	    }
	}
    }

  // Free up memory
  free( x_c );
  free2D( x, *i_N );
  free2D( means, *i_K );
  free3D( icovmats, *i_K, *i_p );
  free2D( delta, *i_N );
}

File: R/qda.R
## qda.R - Quadratic Discriminant Analysis Classifier
##
## by Artem Sokolov

## For all functions:
##
## (N) - Number of samples
## (K) - Number of classes
## (p) - Dimensionality

## Trains a QDA classifier
## x - N x p data matrix of N samples in p dimensions
## y - N x 1 vector of labels
## cov.reg - covariance matrix regularization (towards identity) parameter, must be in [0,1]
##
## Output:
## labels   - vector of class labels
## priors   - K x 1 vector of priors, estimated as fraction of
##		points from each class
## means    - K x p matrix of means approximated from the data
## covmats  - [K x p x p] array of covariance matrices
## icovmats - [K x p x p] array of inverse covariance matrices
## bias	    - K x 1 vector of bias terms for discriminant function computations
QDA.train <- function( x, y, cov.reg=0.0 )
  {
    x <- as.matrix(x)
    clsf <- list()

    ## Argument verification
    if( cov.reg < 0 ) cov.reg <- 0
    if( cov.reg > 1 ) cov.reg <- 1
    
    ## Retrieve labels
    clsf$labels <- unique(y)
    K <- length( clsf$labels )
    N <- nrow(x)
    p <- ncol(x)

    if( SM.getOption( "pure.R" ) == FALSE )
      {
        ## Reindex the labels
        y.i <- rep( 0, length(y));
        for( i in 1:length(y) ) y.i[i] <- which( y[i] == clsf$labels )

        res <- .C( "QDAtrain", as.numeric(x), as.integer( y.i ), as.integer( N ), as.integer( p ),
                  as.integer( K ), as.numeric( cov.reg ),
                  as.integer( SM.getOption( "debug.output" ) ), priors = numeric( K ),
                  means = numeric( K*p ), covmats = numeric( K*p*p ), icovmats = numeric( K*p*p ),
                  bias = numeric( K ), status = integer(1), info = integer(1) )
        
        if( res$info != 0 )
          {
            msg <- paste( "The covariance matrix for class", res$info, "is singular\n" )
            stop( msg )
          }
        else
          {
            res <- list(
                        labels = clsf$labels,
                        priors = res$priors,
                        means = matrix( res$mean, K, p ),
                        covmats = array( res$covmats, c( K, p, p ) ),
                        icovmats = array( res$icovmats, c( K, p, p ) ),
                        bias = res$bias
                        )

            test.na <- function( x ) {any(is.na(c(x)))}
            if( any(unlist(lapply( res, test.na ))) == TRUE )
              stop( "Computation in QDA.train are numerically unstable\n" )
            
            return( res )
          }
      }
    else
      {
        d.cat( 1, "Currently running R version of the QDA.train code\n" )
        
        ## Approximate priors, means, bias
        ##  and covariance matrices
        clsf$priors <- rep( 0, K )
        clsf$means <- matrix( 0, K, p )
        clsf$covmats <- array( 0, c(K,p,p) )
        clsf$icovmats <- array( 0, c(K,p,p) )
        for( k in 1:K )
          {
            mask <- y == clsf$labels[k]
            z <- x[ mask, ,drop=FALSE ]
            clsf$priors[k] <- nrow(z) / N
            clsf$means[k,] <- apply( z, 2, mean )
            clsf$covmats[k,,] <- cov.reg * diag(p) + (1-cov.reg) * cov(z)
            clsf$bias[k] <- log(clsf$priors[k]) - 0.5 * sum( log( svd(clsf$covmats[k,,])$d ) )
            clsf$icovmats[k,,] <- solve(clsf$covmats[k,,])
          }
        
        clsf
      }
  }

## Applies previously trained QDA classifier to classify rows of x
##
## x - N x p data matrix of N samples in p dimensions
## clsf - classifier produced by QDA.train()
##
## Output:
## column vector of labels for each row of x
QDA.test <- function( x, clsf )
  {
    x <- as.matrix(x)

    N <- nrow( x )
    p <- ncol( x )
    K <- length(clsf$bias)

    if( SM.getOption( "pure.R" ) == FALSE )
      {
        if( is.null( clsf$means ) == TRUE )
          stop( "Invalid means field in the classifier" )
        if( is.null( clsf$icovmats ) == TRUE )
          stop( "Invalid icovmats field in the classifier" )
        if( is.null( clsf$bias ) == TRUE )
          stop( "Invalid bias field in the classifier" )
          
        res <- .C( "QDAtest",
                  as.numeric(x), as.numeric( clsf$means ), as.numeric( clsf$icovmats ),
                  as.numeric( clsf$bias ),
                  as.integer( N ), as.integer( p ),
                  as.integer( K ), i = integer( N ) )

        i <- res$i
        }
      else
      {
        deltas <- matrix( 0, nrow(x), K )
        for( i in 1:nrow(x) )
          for( k in 1:K )
            {
              x.c <- x[i,] - clsf$means[k,]
              deltas[i,k] <- -0.5 * t(x.c) %*% clsf$icovmats[k,,] %*% x.c + clsf$bias[k]
            }
        
        i <- apply( deltas, 1, which.max )
      }

    clsf$labels[i]
  }

